### PR TITLE
[BEAM-2430] DoFnRunner for processing a bundle over the Fn API

### PR DESCRIPTION
**Pull request by kennknowles**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

R: @aljoscha this is not super polished, and much of the complexity I thought I would encapsulate really fits better outside the DoFnRunner. But I thought reviewing this and getting it in might be a good way to proceed and share understanding of the Fn API.

### CODE REVIEW

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    public abstract BeamFnApi.Target getTarget();
+
+    public static LogicalEndpoint of(String instructionId, BeamFnApi.Target target) {
+      return new AutoValue_FnDataService_LogicalEndpoint(instructionId, target);
+    }
+  }
+
+  /**
+   * Registers a receiver to be notified upon any incoming
```
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java]: nit (again): unfinished sentence

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      FnDataReceiver<WindowedValue<T>> listener)
+      throws Exception;
+
+  /**
+   * Creates a receiver to which you can write data values and have them sent over this data plane
+   * service.
+   *
+   * <p>The provided coder is used to encode elements on the outbound stream.
+   *
+   * <p>Closing the returned consumer signals the end of the stream.
```
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java]: nit: returned receiver

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * The {@link FnDataService} is able to forward inbound elements to a consumer and is also a
+ * consumer of outbound elements. Callers can register themselves as consumers for inbound elements
+ * or can get a handle for a consumer for outbound elements.
+ */
+public interface FnDataService {
+
+  @AutoValue
+  abstract class LogicalEndpoint {
+    public abstract String getInstructionId();
```
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java]: The instruction id is used to correlate with the instruction that caused the emission (and receiving?) of data?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnApiControlClient.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+  /**
+   * A private view of this class as a {@link StreamObserver} for connecting as a gRPC listener.
+   */
+  private class ResponseStreamObserver implements StreamObserver<BeamFnApi.InstructionResponse> {
+    /**
+     * Processes an incoming {@link BeamFnApi.InstructionResponse} by correlating it with the
+     * corresponding {@link BeamFnApi.InstructionRequest} and completed the future that was returned
```
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnApiControlClient.java]: nit: completes

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnApiControlClientPoolService.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import io.grpc.BindableService;
+import io.grpc.stub.StreamObserver;
+import java.util.concurrent.BlockingQueue;
+import org.apache.beam.fn.v1.BeamFnApi;
+import org.apache.beam.fn.v1.BeamFnControlGrpc;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A Fn API control service which adds incoming SDK harness connections to a pool. */
+public class FnApiControlClientPoolService extends BeamFnControlGrpc.BeamFnControlImplBase
```
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnApiControlClientPoolService.java]: nit: Isn't `BeamFnControlImplBase` already a `BindableService`?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SdkHarnessDoFnRunner.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+/** Processes a bundle by sending it to an SDK harness over the Fn API. */
+public class SdkHarnessDoFnRunner<InputT, OutputT> implements DoFnRunner<InputT, OutputT> {
+
+  private final SdkHarnessClient sdkHarnessClient;
+  private final String processBundleDescriptorId;
+
+  /** null between bundles */
+  @Nullable private SdkHarnessClient.ActiveBundle activeBundle;
+
+  private FnDataReceiver<WindowedValue<InputT>> inputReceiver;
```
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SdkHarnessDoFnRunner.java]: This is not used anywhere.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java (hunk 3 above)
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java]: Yes, specifically it is the ProcessBundleRequest. Added this in javadoc.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java (hunk 1 above)
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java]: Done

File: runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java (hunk 2 above)
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/FnDataService.java]: Done

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SdkHarnessDoFnRunner.java (hunk 6 above)
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SdkHarnessDoFnRunner.java]: Done (at first I thought I would hold it from the constructor, but refactored so the ActiveBundle handles it which is a bit higher level)

**Comment by aljoscha** [comment]
I would be in favour of merging this and evolving as we go (if it has to, at all). What about the `TODOs` in the tests, though?

**Comment by kennknowles** [comment]
Addressed the test TODOs briefly, mostly making clear that these are just smoke tests. I'll merge and move along, as we'll need data service wrappers and factories and whatnot in order to test things meaningfully. And a fleshed-out fake harness will probably help.

**Pull request merged**